### PR TITLE
Save Alignment Data to mount before sync

### PR DESCRIPTION
**Pull request by sywong2000**
Save Alignment Data to mount before sync.
Users can checked the "Save Star Alignment to mount before sync". This would send the star alignment command before every sync operations. Then use the "Mount Model Tool", choose the "Sync" as the platesolve operation. Then at each sync, the alignment data will save to the mount.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#1564: Add Star Alignment option to Rainbow Astro RST-135(E) mount driver](https://issuehunt.io/repos/58881442/issues/1564)
---
</details>

### CODE REVIEW

File: drivers/telescope/rainbow.cpp [hunk 1]
```diff
@@ -230,6 +241,17 @@ bool Rainbow::ISNewSwitch(const char *dev, const char *name, ISState *states, ch
             IDSetSwitch(&HomeSP, nullptr);
             return true;
         }
+        else if (!strcmp(SaveAlignBeforeSyncSP.name, name)) {
+            if (SaveAlignBeforeSyncS[0].s == ISS_ON)
```
Comment (knro) [inline drivers/telescope/rainbow.cpp:245]: No need to do this, just do IUUpdateSwitch(...) then SaveAlignBeforeSyncSP.s = IPS_OK, then setIDSetSwitch(&Save...)

File: drivers/telescope/rainbow.h [hunk 2]
```diff
@@ -160,6 +176,10 @@ class Rainbow : public INDI::Telescope, public INDI::GuiderInterface
         const std::string getSlewErrorString(uint8_t code);
         uint8_t m_SlewErrorCode {0};
 
+        ISwitchVectorProperty SaveAlignBeforeSyncSP;
+        ISwitch SaveAlignBeforeSyncS[1];
```
Comment (knro) [inline drivers/telescope/rainbow.h:180]: I think it's better to make this SaveAlignBeforeSyncS[2] and use INDI_ENABLED and INDI_DISABLED switches. I understand you want to use it as a toggle (which works), but using two switches (Enabled and Disabled) is more *clear* to the end user.

Finally, try to add this switch to the saveConfig(...).

File: drivers/telescope/rainbow.cpp [hunk 3]
```diff
@@ -922,8 +947,17 @@ bool Rainbow::Abort()
...
-    char cmd[DRIVER_LEN] = {0};
 
+    if (SaveAlignBeforeSyncS[0].s == ISS_ON)
+    {
+           StarAlign(ra, dec);
+    }
+    else
+    {
+        LOG_DEBUG("SaveAlignBeforeSyncS is OFF. Not doing anything.");
+    }
```
Comment (knro) [inline drivers/telescope/rainbow.cpp:958]: so starAlign and regular sync can be run after each other? or ONLY one should be used at a time?

**Review by knro — CHANGES_REQUESTED** [review]
Few changes and clarifications.

File: drivers/telescope/rainbow.cpp (hunk 1 above)
Comment (sywong2000) [reply drivers/telescope/rainbow.cpp]: Thanks I have made the changes to my latest push

File: drivers/telescope/rainbow.h (hunk 2 above)
Comment (sywong2000) [reply drivers/telescope/rainbow.h]: Yes definitely better to align to the common UX, I've updated this also

File: drivers/telescope/rainbow.cpp (hunk 3 above)
Comment (sywong2000) [reply drivers/telescope/rainbow.cpp]: yes you're definitely correct, Chris also mentioned this so I changed the command and give the status printout updated a little bit

**Review by knro — COMMENTED** [review]
THanks, looks good but I see many indention and style changes. Are you using **astyle** from INDI?

**Comment by sywong2000** [comment]
I used Qt Creator and I must have hit the Ctrl-I and changed the indents before I commit.. sorry about that.... meanwhile Chris has submitted a version which is very similar. If mine having too much code change it's okay to reject my version and use his version instead. Thank you.

**Comment by knro** [comment]
I don't see any other version submitted?

**Comment by sywong2000** [comment]
Sorry I don't have access to my latest source now, I'm going to submit another version with the astyle later tonight (I'm in GMT +8:00) I'll re-request the review once I've pushed the latest version. Thanks.

**Comment by jimmy-ck** [comment]
@sywong2000: Sorry, I was very busy in the last couple of days. If needed I can create a new PR with my solution. How shall we proceed?

**Comment by sywong2000** [comment]
I just revert a few versions and commit again a more condensed code hopefully this is much better. Thank you.

**Comment by sywong2000** [comment]
> @sywong2000: Sorry, I was very busy in the last couple of days. If needed I can create a new PR with my solution. How shall we proceed?

@jimmy-ck sure, sorry I just noticed this and I has commit another version to revert the unnecessary indents. Please feel free to go ahead if you want to create another PR and overwrite my changes. Thank you.

**Comment by jimmy-ck** [comment]
@sywong2000 Your latest commit looks good. No need for anything else. :)

**Review by jimmy-ck — APPROVED** [review]
Reviewed (not tested) & looks good.

**Comment by knro** [comment]
Thanks! Can this be tested on the actual mount?

**Comment by sywong2000** [comment]
Yes I tested it with my RST-135E before I pushed the code.

**Comment by knro** [comment]
Thanks, let's ask end users now to test from the nightly PPA.

**Comment by jimmy-ck** [comment]
Oh, I think we messed something up here. :(
I completely overlooked that the new property does not get initialised and updated, therefore is not visible in INDI Control Panel. @sywong2000 Did you really compile and test the latest commit on your mount? This not not work. I'll commit a fix.

**Comment by sywong2000** [comment]
@jimmy-ck My fault - I'm terribly sorry I must have messed up the versions when I fix the code indent problem - thanks for catching this and applying the fix.

**Pull request merged**